### DISCUSSION TITLE
Nixos

### DISCUSSION THREAD

**Discussion by zarMarco** [zbm-dev/zfsbootmenu#602] [Feature Requests]
Hi all, tonight I've installed nixos with zfs, but with grub because I didn't found zfsbootmenu for nixos. Is it possible to create a package for this distribution?

**Comment by ahesford** [comment]
This comes up from time to time, but I think NixOS users will have a difficult time making ZBM work because our [assumptions about boot environments](https://docs.zfsbootmenu.org/en/v2.3.x/general/bootenvs-and-you.html) seem at odds with Nix's desire to place things elsewhere. I'm not sure about the specifics, because I've only played with NixOS a bit, and long before I started thinking about ZFS.

As for the packaging itself, that's a question better addressed by Nix people.

**Comment by zarMarco** [comment]
Ok, now I ask on nixos forum

**Comment by hbsnmyj** [comment]
I played around with NixOS, and used a workaround of:
* Copying nix kernels with newer generations (`/nix/store/hash...Linux-version/bzImage`, and corresponding initramfs to /boot
* Manually set kernel command properties to point to the newest nix init (`/nix/store/hash...init`)

I think NixOS will work very well with the general idea of boot environments. we can view each new nix os generation as a bootable kernel version under the same BE, a (kernel, initramfs, kernel arguments) tuple. The problem is ZFSBootMenu cannot recognize it right now. Wondering if there are a suitable place to add some sort of hook/ZFS properties that discovers/captures this information to zbm.

The advantages to the user would be:
* Not need to mess with a separate bootloader for NixOS.
* Supporting native encryption without bpool.
* Rollback/Debug functionality of ZBM is still useful as not necessarily all nix configuration updates are hermetic.

**Reply by almereyda** [reply]
This strategy also seems to be the one applied in [ZFSBootMenu on NixOS - Gridth](https://grid.in.th/2024/12/zfsbootmenu_on_nixos/).
